src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -143,101 +371,179 @@ check if we have not found a block (when block == null
+            {
+                tempRotation = BlockUtils.getRotationFromFacing(blockState.getValue(AbstractBlockHut.FACING));
+            }
+        }
+        else
+        {
+            tempRotation = workOrder.getRotation();
+        }
+
+        loadStructure(workOrder.getStructureName(), tempRotation, pos);
+
+        workOrder.setCleared(false);
+        workOrder.setRequested(false);
+
+        requestMaterials();
     }
 
     /**
-     * Calculates the working position.
-     * <p>
-     * Takes a min distance from width and length.
-     * <p>
-     * Then finds the floor level at that distance and then check if it does contain two air levels.
-     *
-     * @param offset the extra distance to apply away from the building.
-     * @return BlockPos position to work from.
+     * Searches a handy block to substitute a non-solid space which should be guaranteed solid.
+     * @param location the location the block should be at.
+     * @return the Block.
      */
-    private BlockPos getWorkingPosition(final int offset)
+    public abstract Block getSolidSubstitution(BlockPos location);
+
+    /**
+     * Loads the structure given the name, rotation and position.
+     * @param name the name to retrieve  it.
+     * @param rotateTimes number of times to rotate it.
+     * @param position the position to set it.
+     */
+    public void loadStructure(@NotNull final String name, int rotateTimes, BlockPos position)
     {
-        if (offset > MAX_ADDITIONAL_RANGE_TO_BUILD)
+        if(job instanceof AbstractJobStructure)
         {
-            return currentStructure.getCurrentBlockPosition();
+            rotation = rotateTimes;
+            try
+            {
+                 StructureWrapper wrapper = new StructureWrapper(world, name);
+                ((AbstractJobStructure) job).setStructure(wrapper);
+                currentStructure = new Structure(world, wrapper, Structure.Stage.CLEAR);
+            }
+            catch (final IllegalStateException e)
+            {
+                Log.getLogger().warn(String.format("StructureProxy: (%s) does not exist - removing build request", name), e);
+                ((AbstractJobStructure) job).setStructure(null);
+            }
+
+            ((AbstractJobStructure) job).getStructure().rotate(rotateTimes);
+            ((AbstractJobStructure) job).getStructure().setPosition(position);
         }
-        //get length or width either is larger.
-        final int length = currentStructure.getLength();
-        final int width = currentStructure.getWidth();
-        final int distance = Math.max(width, length) + MIN_ADDITIONAL_RANGE_TO_BUILD + offset;
-        @NotNull final EnumFacing[] directions = {EnumFacing.EAST, EnumFacing.WEST, EnumFacing.NORTH, EnumFacing.SOUTH};
+    }
 
-        //then get a solid place with two air spaces above it in any direction.
-        for (final EnumFacing direction : directions)
+    private void requestMaterials()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "requestMaterials" is 15 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)